Continuity ppm port to gpu

edoyango opened the pull request:
Hi Marshall,

Here's the ported continuity_ppm. I wanted to run over it with a finer comb to minimize data transfers in the step_rk2 routine, so opening as draft. Also need to check that the OBC changes that were made in dev/gfdl haven't accidentally been erased during the rebase. At this point most needed data transfers happen in step_rk2, but there's some data transfer happening with OBC which I still want to investigate.

Would it also be helpful for you if i clean up the git history?

Will still be helpful for you to test either way.

PS i accidentally opened a pull request to mom-ocean so please ignore that.

edoyango committed:
zonal_mass_flux: isolate zonal_flux_layer

edoyango committed:
zonal_mass_flux: seperate local_specified_BC block loop

edoyango committed:
zonal_mass_flux: add j dim to tmp vars

edoyango committed:
zonal_mass_flux: seperate visc_rem_max init loops

edoyango committed:
zonal_mass_flux: seperate du_min/max_CFL init loops

edoyango committed:
zonal_mass_flux: separate duhdu/uh_tot_0 init loop

edoyango committed:
zonal_mass_flux: separate du_min/max_CFL aggress_adjust update case (untested)

edoyango committed:
zonal_mass_flux: separate du_min/max_CFL non-aggress_adjust update case

edoyango committed:
zonal_mass_flux: separate du_min/max_CFL non-use_visc_rem update case (untested)

edoyango committed:
zonal_mass_flux: separate du_min/max_CFL non-use_visc_rem update case (untested)

edoyango committed:
zonal_mass_flux: separate du_min/max_CFL 0-clamp loop

edoyango committed:
zonal_mass_flux: separate do_I local_specified_BC init loop (untested)

edoyango committed:
copy zonal_flux_adjust that accepts 2d args

edoyango committed:
zonal_mass_flux move j loop into zonal_flux_adjust copy

edoyango committed:
zonal_flux_adjust_fused: use 2d internal arr

edoyango committed:
zonal_flux_adjusted_fused: separate all loops

edoyango committed:
zonal_mass_flux: separate u/du_cor update (untested)

edoyango committed:
zonal_mass_flux: replicate former present(uhbt) control flow

edoyango committed:
copy set_zonal_BT_cont with 2d args

edoyango committed:
copy set_zonal_BT_cont with 2d args

edoyango committed:
zonal_mass_flux: move j-loop into set_zonal_BT_cont_fused

edoyango committed:
set_zonal_BT_cont_fused: use zonal_flux_adjust_fused

edoyango committed:
set_zonal_BT_cont_fused: separate init loop

edoyango committed:
set_zonal_BT_cont_fused: separate duR/L init loop

edoyango committed:
set_zonal_BT_cont_fused: separate u_0/L/R init loop

edoyango committed:
set_zonal_BT_cont_fused: use zonal_flux_layer_fused

edoyango committed:
set_zonal_BT_cont_fused: separate last 2 loops

edoyango committed:
copy merid_flux_layer that accepts entire arrs

edoyango committed:
merid_flux_layer_fused: separate loops

edoyango committed:
merididional_mass_flux: separate local_specified_BC loop (untested)

edoyango committed:
merididional_mass_flux: separate tmp variable init loops

edoyango committed:
merididional_mass_flux: separate dv_min/max_CFL calc loops

edoyango committed:
merididional_mass_flux: separate dv_min/max_CFL 0 clamp loop

edoyango committed:
merididional_mass_flux: separate simple_OBC_pt init loop (untested)

edoyango committed:
copy meridional_flux_adjust that accepts entire arrs

edoyango committed:
copy meridional_flux_adjust that accepts entire arrs

edoyango committed:
meridional_mass_flux: separate meridional_flux_adjust_fused

edoyango committed:
meridional_mass_flux: move j-loop into meridional_flux_adjust_fused

edoyango committed:
meridional_flux_adjust_fused: separate vh_aux,dvhdv init loops

edoyango committed:
meridional_flux_adjust_fused: 1d->2d tmp arrs

edoyango committed:
meridional_flux_adjust_fused: separate all arrs

edoyango committed:
meridional_mass_flux: separate (d)v_cor assignment loops

edoyango committed:
copy set_merid_BT_cont that accepts entire arrs

edoyango committed:
copy set_merid_BT_cont that accepts entire arrs

edoyango committed:
meridional_mass_flux: move j-loop into set_merid_BT_cont_fused

edoyango committed:
set_merid_BT_cont_fused: use meridional_flux_adjust_fused

edoyango committed:
set_merid_BT_cont_fused: separate tmp var init

edoyango committed:
set_merid_BT_cont_fused: separate short circuit loop

edoyango committed:
set_merid_BT_cont_fused: rm redundant k loop in short circuit

edoyango committed:
set_merid_BT_cont_fused: rm redundant k loop in short circuit

edoyango committed:
set_merid_BT_cont_fused: rm redundant k loop in short circuit

edoyango committed:
set_merid_BT_cont_fused: separate dvL/R init loop

edoyango committed:
set_merid_BT_cont_fused: make remaining tmp arrs 3d

edoyango committed:
set_merid_BT_cont_fused: use merid_flux_layer_fused

edoyango committed:
set_merid_BT_cont_fused: separate last loop

edoyango committed:
meridional_mass_flux: separate any_simple_OBC loop

edoyango committed:
zonal_edge_thickness: move k loop->PPM_reconstruction_x

edoyango committed:
meridional_edge_thickness: move k loop->PPM_reconstruction_y

edoyango committed:
PPM_reconstruction_x/y: move k loop->PPM_limit_pos

edoyango committed:
PPM_reconstruction_x/y: move k loop->PPM_limit_cw84 (untested)

edoyango committed:
zonal_BT_mass_flux: separate all loops (untested)

edoyango committed:
meridional_BT_mass_flux: separate all loops (untested)

edoyango committed:
set_zonal_BT_cont_fused: clean up var defs

edoyango committed:
use SZJB_(G) in do_I dclrns in merid* subroutines

edoyango committed:
set_merid_BT_cont_fused: clean up var defs

edoyango committed:
set_merid/zonal_BT_cont_fused -> set_merid/zonal_BT_cont

edoyango committed:
zonal_flux_adjust_fused: clean up var defs

edoyango committed:
zonal_flux_adjust_fused -> zonal_flux_adjust

edoyango committed:
zonal_flux_adjust_fused -> zonal_flux_adjust

edoyango committed:
zonal_flux_layer_fused -> zonal_flux_layer

edoyango committed:
meridional_flux_adjust_fused: clean up var defs

edoyango committed:
meridional_flux_adjust_fused -> meridional_flux_adjust

edoyango committed:
meridional_flux_adjust_fused -> meridional_flux_adjust

edoyango committed:
merid_flux_layer_fused: clean up var defs

edoyango committed:
merid_flux_layer_fused -> merid_flux_layer

edoyango committed:
fix call merid/zonal_flux_layer line conts

edoyango committed:
zonal_flux_layer: move GV in var dclrn

edoyango committed:
zonal_mass_flux: use visc_rem_u_tmp

edoyango committed:
meridional_mass_flux: use visc_rem_v_tmp

edoyango committed:
zonal/meridional_mass_flux: remove old visc_rem vars

edoyango committed:
remove redundant kloop

edoyango committed:
remove problematic omp directives

edoyango committed:
port PPM_limit_pos/CW84, PPM_reconstruction_x, zonal_edge_thickness

edoyango committed:
PPM_reconstruction_x: add enter/exit data stmts

edoyango committed:
continuity_zonal_convergence: save loop range for porting convenience

edoyango committed:
port continuity_zonal_convergence

edoyango committed:
zonal_mass_flux: array init -> loop init

edoyango committed:
zonal_mass_flux: port init loops

edoyango committed:
port zonal_flux_layer

edoyango committed:
zonal_flux_layer: add enter/exit data stmts

edoyango committed:
port zonal_flux_adjust

edoyango committed:
port set_zonal_bt_cont

edoyango committed:
zonal_flux_adjust: add more arrs in enter/exit data

edoyango committed:
set_zonal_BT_cont: add enter/exit data stmts

edoyango committed:
port zonal_flux_thickness loops

edoyango committed:
zonal_flux_thickness add enter/exit data stmts

edoyango committed:
zonal_mass_flux prepare obc for porting

edoyango committed:
zonal_mass_flux: merge any_somple_OBC loops

edoyango committed:
zonal_mass_flux port main loops

edoyango committed:
zonal_mass_flux port OBC loops

edoyango committed:
zonal_mass_flux: add enter/exit data stmts

edoyango committed:
continuity_ppm: add initial enter/exit data stmts

edoyango committed:
port PPM_reconstruction_y, meridional_edge_thickness

edoyango committed:
port continuity_merdional_convergence

edoyango committed:
ppm_reconstruction_y: add enter/exit data stmts

edoyango committed:
port merid_flux_layer

edoyango committed:
meridional_flux_adjust: port loops

edoyango committed:
meridional_flux_adjust: add enter/exit data stmts

edoyango committed:
meridional_mass_flux: port core loops

edoyango committed:
meridional_mass_flux: port OBC loops (untested)

edoyango committed:
meridional_flux_thickness: port core loops

edoyango committed:
meridional_flux_thickness: attempt port OBC loops (untested)

edoyango committed:
meridional_flux_thickness: add enter/exit data stmts

edoyango committed:
port set_merid_BT_cont loops

edoyango committed:
set_merid_bt_cont: add enter/exit data stmts

edoyango committed:
meridional_mass_flux: add enter/exit data stmts

edoyango committed:
zonal/meridional_mass_flux: add missing vars in enter/exit stmts

edoyango committed:
continuity_PPM: complete enter/exit data stmts

edoyango committed:
*_edge_thickness: do concurrent

edoyango committed:
zonal_mass_flux: do concurrent-ify

edoyango committed:
set_zonal_BT_cont: do concurrent-ify

edoyango committed:
zonal_flux_adjust: do concurrent-ify

edoyango committed:
zonal_flux_layer: do concurrent

edoyango committed:
zonal_flux_thickness: do concurrent

edoyango committed:
continuity_zonal_convergence: do concurrent

edoyango committed:
meridional_mass_flux: do concurrent

edoyango committed:
set_merid_bt_cont: do concurrent

edoyango committed:
meridional_flux_adjust: do concurrent

edoyango committed:
merid_flux_layer: do concurrent

edoyango committed:
meridional_flux_thickness: do concurrent

edoyango committed:
continuity_merdional_convergence: do concurrent

edoyango committed:
formatting

edoyango committed:
continuity_PPM: update LB

Commit:
meridional/zonal_flux_adjust: separate ij-reduction

edoyango committed:
zonal_flux_adjust: a couple jki loops

edoyango committed:
set_zonal_bt_cont: some jki loops

edoyango committed:
zonal_mass_flux: some jki loops

edoyango committed:
optimise loops by using scalar zonal_flux_layer

edoyango committed:
zonal_flux_adjust: duhdu -> scalar

edoyango committed:
elemental zonal_flux_layere and separated OBC

edoyango committed:
zonal_flux_adjust: back to jki

edoyango committed:
zonal_flux_thickness: improve cpu perf a bit

edoyango committed:
zonal_flux_adjust: add some comments and guard early exit

edoyango committed:
zonal_flux_adjust: use omp target loop for private arrs

edoyango committed:
set_zonal_BT_cont: use omp target loop for private arrs

edoyango committed:
set_zonal_BT_cont: do conc jki loop

edoyango committed:
zonal_mass_flux: rm useless do_i init

edoyango committed:
zonal_flux_layere: precalc g_dy_Cu*por_face_areaU

edoyango committed:
zonal_flux_layere: precalc dh

edoyango committed:
zonal_flux_thickness: precalc dh

edoyango committed:
zonal_flux_adjust: make uhbt optional

edoyango committed:
zonal_flux_thickness: assign to outputs directly

edoyango committed:
mv zonal_flux_adjust from set_zonal_bt_cont->zonal_mass_flux

edoyango committed:
zonal_mass_flux: reuse visc_rem_u_tmp more

edoyango committed:
zonal_mass_flux: remove redundant if stmt

edoyango committed:
zonal_mass_flux: force inline zonal_flux_layere

edoyango committed:
zonal_flux_layere: inlinable gfortran -O3 but slower for ifort

edoyango committed:
zonal_flux_layere: make a bit "smaller"

edoyango committed:
zonal_mass_flux: move present(uhbt) or set_BT_cont to new subroutine

edoyango committed:
Revert "zonal_flux_layere: make a bit "smaller""

This reverts commit 316152eb3cf515c4179f8ceb738f9d259233915c.

edoyango committed:
Revert "zonal_flux_layere: inlinable gfortran -O3 but slower for ifort"

This reverts commit 51ee896c54b4a349acb2ffe70489dc76558864b1.

edoyango committed:
fix gcc line truncation

edoyango committed:
pass doxygen tests

edoyango committed:
remove forceinline dirs

edoyango committed:
attempt document vars

edoyango committed:
a couple long lines

edoyango committed:
last trailing space

edoyango committed:
meridional_mass_flux: use zonal_flux_layere

edoyango committed:
zonal_flux_layere_OBC: make elemental

edoyango committed:
meridional_mass_flux: move dv_min/max_CFL calc into j-loop

edoyango committed:
meridional_mass_flux: move big chunk into separate subroutine

edoyango committed:
present_vhbt_or_set_bt_cont: merge couple of loops

edoyango committed:
present_vhbt_or_set_bt_cont: merge couple of loops

edoyango committed:
meridional_flux_thickness: cpu optimize a bit

edoyango committed:
meridional_flux_thickness: cpu optimize a bit

edoyango committed:
meridional_flux_adjust: back to jki

edoyango committed:
meridional_flux_adjust: back to jki

edoyango committed:
set_merid_BT_cont: pull out meridional_flux_adjst

edoyango committed:
set_merid_BT_cont: pull out meridional_flux_adjst

edoyango committed:
set_meridional_BT_cont: optimize cpu

edoyango committed:
set_meridional_BT_cont: optimize cpu

edoyango committed:
rm remaining merid/zonal_flux_layer

edoyango committed:
rm remaining merid/zonal_flux_layer

edoyango committed:
zonal_flux_layere: improve naming a bit

edoyango committed:
zonal_flux_layere: improve naming a bit

edoyango committed:
zonal_flux_layere_OBC: improve naming a bit

edoyango committed:
zonal_flux_layere_OBC: improve naming a bit

edoyango committed:
improve flux_elem line wraps

edoyango committed:
improve flux_elem line wraps

edoyango committed:
optimize data transfers a bit

edoyango committed:
optimize data transfers a bit

edoyango committed:
pass elem not arr

edoyango committed:
pass elem not arr

edoyango committed:
add some missing documentation

edoyango committed:
add some missing documentation

edoyango committed:
fix trailing spaces and missing var docs

edoyango committed:
fix trailing spaces and missing var docs

edoyango committed:
add last param docs

edoyango committed:
add last param docs

edoyango committed:
meridional_flux_adjust: fix fpe err

edoyango committed:
meridional_flux_adjust: fix fpe err

edoyango committed:
fix another fpe

edoyango committed:
fix another fpe

edoyango committed:
cleanup args for new helper subroutines

edoyango committed:
cleanup args for new helper subroutines

edoyango committed:
clean up enter/exit data

edoyango committed:
clean up enter/exit data

edoyango committed:
fix passing h twice

edoyango committed:
fix passing h twice

edoyango committed:
meridional_mass_flux: do concurrent

edoyango committed:
meridional_mass_flux: do concurrent

edoyango committed:
zonal_flux_adjust: use a few 3d tmp arrays to mirror meridional_flux_adjust

edoyango committed:
zonal_flux_adjust: use a few 3d tmp arrays to mirror meridional_flux_adjust

edoyango committed:
move target update out of continuity

edoyango committed:
move target update out of continuity

edoyango committed:
initialize pbv%por_face_area[U/V] on GPU

edoyango committed:
initialize pbv%por_face_area[U/V] on GPU

edoyango committed:
cleanup some transfers from continuity_PPM

edoyango committed:
cleanup some transfers from continuity_PPM

marshallward commented:
Don't worry about the commit history, it all needs a reset before moving into dev/gfdl anyway.  There are conflicts though which you might want to look over.

I'll try to have a proper look at this by Monday, and we can talk it over next meeting if needed.

edoyango committed:
clean up a few minor things

edoyango committed:
clean up a few minor things

edoyango committed:
zonal/meridional_flux_adjust: use scalar u/v_new

edoyango committed:
zonal/meridional_flux_adjust: use scalar u/v_new

edoyango committed:
declare vp,up,h_tmp on gpu

edoyango committed:
declare vp,up,h_tmp on gpu

edoyango committed:
remove h update

edoyango committed:
remove h update

edoyango commented:
> There are conflicts though which you might want to look over.

Which conflicts are these? Github is saying no conflicts.

Can you see if the latest commits work for you? Jorge had a try on his V100 cluster and got errors and compute-sanitizer complained pretty bad, but my it works fine on my laptop without complaints from compute-sanitizer.

Ok we managed to discover that including `-g` with `-O0` causes errors in this branch. `-g` has to be excluded for things to work correctly.

marshallward commented:
> Which conflicts are these? Github is saying no conflicts.

Sorry about that, they were rebase conflicts, not merge conflicts.  We can probably handle these.

I did finally get a chance to test it out, and `eta` went unstable.

```
WARNING: btstep: eta has dropped below bathyT:  -1.9422290187396845E+02 vs.  -1.7318475656644790E+02 at   2.1750E+01  3.9250E+01     44     19


WARNING: btstep: eta has dropped below bathyT:  -1.9422290187396845E+02 vs.  -1.7318475656644790E+02 at   2.1750E+01  3.9250E+01     44     19


WARNING: btstep: eta has dropped below bathyT:  -4.8188594703127075E+02 vs.  -4.5788624238102386E+02 at   2.0750E+01  4.0250E+01     42     21

(and so on...)
```
Most likely some array or value is not being copied to or from the GPU.

I'll try to look more into it.

marshallward commented:
I see this diff in the `btcalc()` output, i believe it's the third call of `btcalc().

```
   501 u-point: mean=   4.8863636363636365E-01 min=   0.0000000000000000E+00 max=   1.0000000000|   501 u-point: mean=   4.8863636363636365E-01 min=   0.0000000000000000E+00 max=   1.0000000000
   502 u-point: c=    108126 W=    108126 u btcalc frhat[uv]                                    |   502 u-point: c=    107816 W=    107816 u btcalc frhat[uv]                                    
   503 v-point: mean=   4.8749999999999999E-01 min=   0.0000000000000000E+00 max=   1.0000000000|   503 v-point: mean=   4.8749999999999999E-01 min=   0.0000000000000000E+00 max=   1.0000000000
   504 v-point: c=    105966 S=    105966 v btcalc frhat[uv]                                    |   504 v-point: c=    106990 S=    106990 v btcalc frhat[uv]                                    
   505 u-point: mean=   6.0131543159809280E+02 min=   0.0000000000000000E+00 max=   1.0000003825|   505 u-point: mean=   6.0131554906284782E+02 min=   0.0000000000000000E+00 max=   1.0000009065
   506 u-point: c=    115172 W=    115172 u btcalc h_[uv]                                       |   506 u-point: c=    115229 W=    115229 u btcalc h_[uv]                                       
   507 v-point: mean=   5.9322100180252733E+02 min=   0.0000000000000000E+00 max=   1.0000002211|   507 v-point: mean=   5.9322100184810483E+02 min=   0.0000000000000000E+00 max=   1.0000001633
   508 v-point: c=    105183 S=    105183 v btcalc h_[uv]                                       |   508 v-point: c=    114320 S=    114320 v btcalc h_[uv]                                       
```
But this PR does not touch `MOM_barotropic.F90`, so perhaps something happened with `CS%BT_cont%h_[uv]`

(I am running v24.11 on an H-100).

edoyango commented:
| I did finally get a chance to test it out, and eta went unstable.

This happened to both jorge and I if compiled with `-g` (with no optimisation flags) or `-O0 -g`. Are you compiling with these flags?

edoyango commented:
update: changing `flux_elem` from `elemental` to `pure` didn't change the unstable eta.

Seems like getting rid of `-g` is the only option.

edoyango committed:
continuity_PPM: minimize mapping stmts

edoyango committed:
continuity_PPM: minimize mapping stmts

edoyango committed:
zonal_flux_adjust: minimize mapping stmts

edoyango committed:
zonal_flux_adjust: minimize mapping stmts

edoyango committed:
set_zonal_bt_cont: minimize mapping stmts

edoyango committed:
set_zonal_bt_cont: minimize mapping stmts

edoyango committed:
merional_flux_adjust: minimize mapping stmts

edoyango committed:
merional_flux_adjust: minimize mapping stmts

edoyango committed:
set_merid_bt_cont: minimize mapping stmts

edoyango committed:
set_merid_bt_cont: minimize mapping stmts

edoyango committed:
zonal/meridional_flux_adjust: tmp vars duhdu/dvhdv 3d -> scalar

edoyango committed:
zonal/meridional_flux_adjust: tmp vars duhdu/dvhdv 3d -> scalar

marshallward commented:
ifort 2023 (our system build) is choking on some of these newer constructs.

1. It doesn't understand `!$omp target loop`
   ```
   $MOM6/src/core/MOM_continuity_PPM.F90(1338): error #5082: Syntax error, found 'LOOP' when expecting one of: DATA EXIT ENTER VARIANT TEAMS SIMD PRIVATE ALLOCATE FIRSTPRIVATE MAP ...
   !$omp target loop private(uh_err, uh_err_best, duhdu_tot, du_min, du_max, do_I) &
   ```

2. It doesn't like variables in both `private()` and `map()`:
   ```
   $MOM6/src/core/MOM_continuity_PPM.F90(1339): error #8327: Duplicate variable in list.   [DO_I]
   !$omp target loop private(vh_err, vh_err_best, dvhdv_tot, dv_min, dv_max, do_I) &                                                                                                                       
   !$omp   map(alloc: vh_aux, v_new, vh_err, vh_err_best, dvhdv_tot, dv_min, dv_max, do_I)
   ```
There could be more, those are the errors so far.

I don't know how vital it is to solve these issues.  It may be expecting too much for any version of `ifort` to understand OpenMP offloading.  We could just disable openmp testing.  But we should check with the consortium if there's no way to avoid these unsupported constructs.

marshallward commented:
I tried running without `-g` but it still seems to be crashing (negative `eta`).

FWIW these are my flags: `-O0 -mp=gpu -stdpar=gpu -gpu=mem:separate,debug -Mnofma`

edoyango commented:
> FWIW these are my flags: -O0 -mp=gpu -stdpar=gpu -gpu=mem:separate,debug -Mnofma

Can you try without `-gpu=debug`? This gives me `eta` errors too.

> It doesn't understand `!$omp target loop`

This will probably be a problem with AMD too. Alternatively, we can use `do concurrent`, but the arrays that are `private` will have to be promoted to 2D since NVIDIA `do concurrent` doesn't accept `local` arrays. I'm happy to do that, but was conscious of your goal of minimizing local array sizes. Should I leave it as-is or convert do concurrent with 2d tmp arrays?

> It doesn't like variables in both `private()` and `map()`:

That's fine, I can add a separate `!$omp target enter/exit data`.

marshallward commented:
> Can you try without -gpu=debug? This gives me eta errors too.

OK there is goes. :upside_down_face:  At least I can replicate it.

edoyango commented:
@marshallward Some CPU timings for you to stew on when you get back!

In summary, 
* the ported code in this branch has poorer memory access patterns when the problem size is large enough. 
    * this code is **faster than original for `double_gyre` **
    * this code is **slower than original for `benchmark`** in 360x180 or 180x90 for NIGLOBALxNJGLOBAL
* **With do concurrent, `ifort` with light optimisations are roughly within 10% slower of the original code** for NCI's cascade lake and sapphirerapids nodes
* **with nested do instead of do concurrent, run times are up to 15% worse than the original**
* Given these results, merging this code into dev/gfdl, but without do concurrents will result in a noticeable performance hit to CPU. Can discuss more when you're back.

below are some timings comparison for this branch on CPU with do concurrent (`newdoc`), and without do concurrent (`new`). This is compared to dev/gfdl (`base`). Tests are with benchmark case. `360x180` and `180x90` are the `NIGLOBAL, NJGLOBAL` values respectively. Times used are the `continuity equation` time reported in stdout.

```
─ Results for: 360x180 | ifort | O2-g-mavx2 | normalsr ─
  Baseline 'base' (Avg Time: 26.3026s)
  ❌ new:         Slower               (+14.30% avg time)
  ❌ newdoc:      Slower               (+11.02% avg time)

─ Results for: 180x90 | ifort | O2-g-mavx2 | normalsr ─
  Baseline 'base' (Avg Time: 7.0465s)
  ❌ new:         Slower               (+8.10% avg time)
  ❌ newdoc:      Slower               (+2.93% avg time)
```
So main takeaways are that the do concurrent version of this branch is close to within 10% of dev/gfdl on CPU with ifort and light optimisations on both sapphirerapids and cascadelake nodes at NCI. But, `do concurrent` is faster than nested `do` by a noticeable amount.

Annoyingly, `ifx` is almost 100% slower with the same compile options than the base code also compiled with `ifx`. Although the difference is only ~6% when aggressive optimisations are turned on `-O3 -flto -march=sapphirerapids -mtune=sapphirerapids`

`double_gyre` shows some speedup which says (to me) the memory access patterns in the ported code are worse.

```
─ Results for: double_gyre | ifort | O2-g-mavx2 | normalsr ─
  Baseline 'base' (Avg Time: 2.3534s)
  ❌ new:         Slower               (+0.06% avg time)
  ⚠️ newdoc:      Potentially Faster  (-9.03% avg time)
```

edoyango committed:
separate alloc of private variables for gcc

edoyango committed:
separate alloc of private variables for gcc

edoyango committed:
u/vh_aux: 3d->2d

edoyango committed:
u/vh_aux: 3d->2d

edoyango committed:
target teams loop recognizable by amdflang

edoyango committed:
target teams loop recognizable by amdflang

edoyango commented:
> It doesn't understand !$omp target loop

As per jorge's suggestions, I've switched to `teams loop` instead of just `loop` which `ifort` understands (and also `flang` and `gcc`).

> It doesn't like variables in both private() and map():

Moved mappings to separate enter/exit statements.

marshallward commented:
I just tested this out (with `-g`!) and it seems to not only run but also give the same answers as CPU.  Was there some change here that fixed the problem?

edoyango commented:
Oh I hadn't noticed! The only thing I changed was demoting a 3d tmp array to 2d. The array was 3d because of the issues I was having, but they went away after the rebase...

marshallward commented:
In my enthusiasm for #41 it looks like I may have broken this PR :(.  I am now getting answer changes.

AFAIK the change is in the first continuity call
https://github.com/edoyango/MOM6/blob/d2b7fa4417ebdc5ca6c5cf7050c356ed8ef0830e/src/core/MOM_dynamics_split_RK2.F90#L803-L807

I made a little progress by removing the `map(to: h)`/`map(release: h)` inside `continuity_PPM()`, but I still see a diff in the `uh`, `vh` output.

(Should have mentioned, this is after a `dev/gpu` rebase)

marshallward commented:
Nevermind!  Moving the allocation of `CS%continuity_CSp` outside of `continuity_init` seems to have fixed it.  I am getting the old answers again.

This is my mess, so I'll send a PR to this branch.

marshallward commented:
I realize now that a patch won't make much sense until this branch is rebased against dev/gpu...

marshallward committed:
Continuity CS outside of init

This moves the Continuity CS to the dycore init function.  For some
reason, this avoids an answer change with CPU.  (Possibly because alloc
inside of a function doesn't quite match the CS outside of it?)

A few minor data transfers are also added to fix up differences in the
chksum log output.

marshallward committed:
Continuity CS outside of init

This moves the Continuity CS to the dycore init function.  For some
reason, this avoids an answer change with CPU.  (Possibly because alloc
inside of a function doesn't quite match the CS outside of it?)

A few minor data transfers are also added to fix up differences in the
chksum log output.

marshallward commented:
Here is my working commit: 53588eca91f0417cfb3c227caa540f1ee4d176db

https://github.com/marshallward/MOM6/tree/cont_ppm_rebase

edoyango commented:
> Here is my working commit: [53588ec](https://github.com/marshallward/MOM6/commit/53588eca91f0417cfb3c227caa540f1ee4d176db)
> 
> https://github.com/marshallward/MOM6/tree/cont_ppm_rebase

Works for me - I've updated my branch to use your branch

marshallward commented:
~I still can't figure out why GitHub is reporting rebase conflicts here.  I have not yet seen any rebase issues, and you just applied a full rebase to dev/gpu.  Any idea what this could be?~

Nevermind, we can just squash-merge it.

marshallward committed:
Merge branch 'dev/gpu' into continuity_ppm-gpu

marshallward commented:
@edoyango Do you know what might be causing the OpenMP CI fail?  Only `[uv]hbt` seems to fail, but I don't see anything exceptional about it.

marshallward commented:
I tested this on my machine and it passed.  MacOS must be hitting something particular here.  Perhaps there is some bad interaction with old and new OpenMP directives.

marshallward commented:
I think I see the issue.  Most likely it's just checksums.  As one is fixed, the next reveals itself.  And since the log is a `tail`, we don't see the fixed ones.

Previously all the `update()` around `btstep` were inadvertently sorting out these checksums, but now they need to be manually downloaded.

I'll try to work through these.  I will also update my compiler (although that can take many hours to build.)

marshallward commented:
I made a little headway but I don't think it's as simple as checksum cleanup.  There seems to be some diff that I'm missing.  As for why changes to continuity are introducing NaNs in barotropic, I am at a loss.

I'll continue looking, but I might need some help.

marshallward commented:
I've narrowed down the issue to the `!$omp target teams loop` directives.  All four appear to cause a failure.  The diagnosis is unfortunately statistical, but this is roughly what I see.

* Over 15 tests passed hen all `!$omp target teams loop` directives are disabled.
* About 1 in 3 tests fail when at least one is enabled.
* The data transfer nodes do not cause a failure.  (But I won't yet speculate if they affect the behavior of the `teams loop` directives.)

As we've discussed, the issue is not unique to GNU.  I can see the same problem with Nvidia.  And reproducibility is a challenge.  I can reproduce on three platforms (including GitHub Actions), but @edoyango cannot as yet reproduce the problem.

@edoyango Would the nested do-concurrent loop work here?  It would effectively disable any OpenMP threading, but perhaps that's not a bad thing?

marshallward commented:
The issue was missing `local()` specifiers of the `do concurrent` inside of the `!$omp target teams loop` do-loops.

And yes, it was me who suggested removing them!  But with the new `DO_LOCALITY()` macro, we can add those back without breaking the compilers without full do concurrent support.

I've tested the CPU builds, but I'll do another GPU test to make sure everything works.  If so, then I can merge this in and apply the `local()` patch directly.

marshallward committed:
Continuity: Add locality to do concurrent

Do concurrent inside of !$omp target teams loop seems to fail standard
openmp tests if locality is not correctly set.

This patch adds the correct locality to the four `!$omp target teams
loop` directives.

The domore argument has also been removed, and replaced with a
`.not.any(do_i(:))` test.

marshallward commented:
I updated (and rebased) the `cont_ppm_rebase` branch, which seems to fix these issues.  Ultimately, I think the problem was do concurrent locality specifiers.

I think the message is that we probably ought to go back and fix up the locality of these directives before merging to main.

https://github.com/marshallward/MOM6/tree/cont_ppm_rebase

marshallward commented:
There was a rather serious performance bug in a previous version of the branch, but I've updated and rebased it and the issue seems to be resolved now.

edoyango committed:
Merge remote-tracking branch 'origin/cont_ppm_rebase' into continuity_ppm-gpu

edoyango commented:
I've merged your branch into this one - I ran it against benchmark and double gyre and didn't see any answer changes or performance degradation (maybe cpu got a little bit faster actually - perhaps due to less branching because of domore, but idk). 
I'll need you to teach me how to look at the variable checksums again so I can use it as an additional check.

marshallward commented:
I think this is probably good to take in, so that we can turn our efforts to vert friction and the dycore loop.

We can follow up about the chksum diffs, but if you turn on `DEBUG = True` and save the stderr output, then you should get the idea.

marshallward commented:
I squash-merged, hope it's ok
